Adopt capsules for QUIC-aware proxying (#51)
tfpauly/quic-proxy: draft-pauly-masque-quic-proxy.md

---

> title: QUIC-Aware Proxying Using HTTP

Author: @LPardue the quic-awareness works for all versions of HTTP, it's just forwarding mode that is HTTP/3 only.

> title: QUIC-Aware Proxying Using HTTP

Reviewer: ah fair point. if you change the final word to `HTTP` the acronym still works ;)

--- abstract

This document defines an extension to UDP Proxying over HTTP
that adds specific optimizations for proxied QUIC connections. This extension
allows a proxy to reuse UDP 4-tuples for multiple connections. It also defines a
mode of proxying in which QUIC short header packets can be forwarded through the

> proxy rather than being re-encapsulated and re-encrypted.

Reviewer: It's not until mid-way through the intro does it become clear this only works for HTTP/3 connections.

Author: Clarified this in the abstract

Both clients and proxies can unilaterally choose to disable forwarded mode for
any client <-> target connection.

The forwarding mode of this extension is only defined for HTTP/3

> {{!HTTP3=I-D.ietf-quic-http}} and not any earlier versions of HTTP.

Reviewer 2: Should we also note that this doesn't work in the presence of intermediaries between the client and proxy?

Author: Added text for that

this way allows the proxy to reuse server-facing sockets for multiple
connections and support the forwarding mode of proxying.

QUIC packets can be either tunnelled within an HTTP proxy connection using

> HTTP Datagram frames {{!HTTP-DGRAM=I-D.schinazi-quic-h3-datagram}}, or be forwarded

Reviewer: it's adopted now :)

```suggestion
HTTP Datagram frames {{!HTTP-DGRAM=I-D.masque-h3-datagram}}, or be forwarded
```

the proxy needs to maintain mappings between several items. There are three
required unidirectional mappings, described below.

> ## Quarter Stream ID Mapping

Author: Fixed

to signal the addition and removal of client and server connection IDs.

Note that these capsules do not register contexts or define a new HTTP Datagram
Format. By default, QUIC packets are encoded using HTTP Datagrams with the
UDP_PAYLOAD HTTP Datagram Format Type as defined in {{CONNECT-UDP}}.

The capsules used for QUIC-aware proxying allow a client to register connection
IDs with the proxy, and for the proxy to acknowledge or reject the connection
ID mappings.

## Register and Acknowlege Connection ID Capsule Types

The REGISTER_CLIENT_CID and REGISTER_SERVER_CID capsule types (see 
{{iana-capsule-types}} for the capsule type values) allow a client to inform
the proxy about a new Client Connection ID or a new Server Connection ID,
respectively.

The ACK_CLIENT_CID and ACK_SERVER_CID capsule types (see {{iana-capsule-types}}
for the capsule type values) are sent by the proxy to the client to indicate
that a mapping was successfully created for a registered connection ID.

Register and Ack Connection ID capsule types share the same format:

~~~
Connection ID Capsule {
  Type (i) = 0xffe100..0xffe103,
  Length (i),

>   Connection ID Length (8),

Author: Sounds good, removed

  Connection ID (0..160),
}
~~~
{: #fig-capsule-cid title="Connection ID Capsule Format"}

Connection ID Length:
: An 8-bit unsigned integer containing the length of the connection ID.
Values less than 1 and greater than 20 are invalid.

Connection ID:
: A connection ID of the specified length.

## Close Connection ID Capsule Types {#close-capsule}

The CLOSE_CLIENT_CID and CLOSE_SERVER_CID capsule types (see 
{{iana-capsule-types}} for the capsule type values) allow either a client
or a proxy to remove a mapping for a connection ID. These capsules share
the following format:

~~~
Connection ID Capsule {
  Type (i) = 0xffe104..0xffe105,
  Length (i),
  Connection ID Length (8),
  Connection ID (0..160),
  Close Code (i),
  Close Details (..),
}
~~~
{: #fig-capsule-close-cid title="Close Connection ID Capsule Format"}

> Close Code:

Reviewer 2: Calling these close codes conflicts with HTTP-DGRAM and is going to be confusing. How about CID Close Code? Especially renaming the prose in the rest of the doc would help.

Author: Removed the close codes

: The close code allows an endpoint to provide additional information as
to why a connection ID mapping was closed. The codes are defined in
{{iana-close-codes}}.

Close Details:
: This is meant for debugging purposes. It consists of a human-readable
string encoded in UTF-8.¶

Connection ID Length:
: An 8-bit unsigned integer containing the length of the connection ID.
Values less than 1 and greater than 20 are invalid.

Connection ID:
: A connection ID of the specified length.

# Client Request Behavior {#request}

A client initiates UDP proxying via an extended CONNECT request as defined
in {{CONNECT-UDP}}. It then sends a REGISTER_CLIENT_CID capsule whenever it

> advertises a new Client Connection ID to the target, and a REGISTER_SERVER_CID

Author: We do have that already:

```
Prior to informing the target of a new chosen client connection ID, the client
MUST send a REGISTER_CLIENT_CID capsule request containing the new Client
Connection ID.
```

capsule when it has received a new Server Connection ID for the target.

## New Proxied Connection Setup

> To initiate QUIC-aware proxying, the client sends a REGISTER_CLIENT_CID

Author: Added a header and explained its usage!

any Client Connection ID that has been accepted by the proxy. The client uses
the received Connection ID to determine if a packet was originated by the
proxy, or merely forwarded from the target.

## Opting Out of Forwarded Mode

The use of forwarded mode is initiated by the client sending a request with the
REGISTER_SERVER_CID capsule and sending at least one forwarded packet to the
proxy. A client that does not wish to accept forwarded packets from the proxy
when communicating with a specific target can simply not start forwarding
packets to the proxy.

# Proxy Response Behavior {#response}

Upon receipt of a CONNECT request that includes the "Proxy-QUIC" header,
the proxy indicates to the client that it supports QUIC-aware proxying

> by including the same "Proxy-QUIC" header with a value of "?1" in a
> successful response.

Reviewer: no need to state it's the same if you provide the value to send

```suggestion
by including a "Proxy-QUIC" header with a value of "?1" in a
successful response.
```

Upon receipt of a REGISTER_CLIENT_CID or REGISTER_SERVER_CID capsule,
the proxy validates the registration, tries to establish the appropriate
mappings as described in {{mappings}}.

The proxy MUST reply to each REGISTER_CLIENT_CID capsule with either
an ACK_CLIENT_CID or CLOSE_CLIENT_CID capsule containing the
Connection ID that was in the registration capsule.

Similarly, the proxy MUST reply to each REGISTER_SERVER_CID capsule with 
either an ACK_SERVER_CID or CLOSE_SERVER_CID capsule containing the
Connection ID that was in the registration capsule.

The proxy then determines the server-facing socket to associate with the
client's request. This will generally involve performing a DNS lookup for

> the hostname in the CONNECT request authority, or finding an existing server-facing

Reviewer: It's not the request authority now, it's the path

Author: ```suggestion
the target hostname in the CONNECT request, or finding an existing server-facing
```

If a proxy reuses server-facing sockets, it SHOULD store which authorities
(which could be a domain name or IP address literal) are being accessed over a
particular server-facing socket so it can avoid performing a new DNS query and
potentially choosing a different server IP address which could map to a
different server.

Server-facing sockets MUST NOT be reused across QUIC and non-QUIC UDP proxy
requests, since it might not be possible to correctly demultiplex or direct
the traffic. Any packets received on a server-facing socket used for proxying
QUIC that does not correspond to a known Connection ID MUST be dropped.

When the proxy recieves a REGISTER_CLIENT_CID capsule, it is receiving a
request to be able to route traffic back to the client using that Connection ID.
If the pair of this Client Connection ID and the selected server-facing socket
does not create a conflict, the proxy creates the mapping and responds with a
ACK_CLIENT_CID capsule. After this point, any packets received by the proxy from the
server-facing socket that match the Client Connection ID can to be sent to the
client. The proxy MUST use tunnelled mode (HTTP Datagram frames) on the correct

> datagram context for any long header packets. The proxy SHOULD forward directly to

Reviewer: is the use of context here correct? The example uses REGISTER_NO_CONTEXT

Reviewer 2: I'd just skip mention of datagram contexts altogether in this sentence. "Use the correct thing" isn't particularly useful

Author: Good call! Fixed up this sentence

the client for any matching short header packets, but MAY tunnel them in
HTTP Datagram frames. If the pair is not unique, or the proxy chooses not to support
zero-length Client Connection IDs, the proxy responds with a CLOSE_CLIENT_CID

> capsule. If this occurs on the first request for a given HTTP request, the

Reviewer: "first request for a given HTTP request" reads a bit wonky even if it is accurate

the client for any matching short header packets, but MAY tunnel them in
DATAGRAM frames. If the pair is not unique, or the proxy chooses not to support
zero-length Client Connection IDs, the proxy responds with a CLOSE_CLIENT_CID
capsule with a CONFLICT close code. If this occurs on the first request for a
given HTTP request, the proxy removes any mapping for that HTTP request.

When the proxy recieves a REGISTER_SERVER_CID capsule, it is receiving a
request to allow the client to forward packets to the target. If the pair of
this Server Connection ID and the client-facing socket on which the request was
received does not create a conflict, the proxy creates the mapping and responds
with a ACK_SERVER_CID capsule. Once the successful response is sent, the proxy will
forward any short header packets received on the client-facing socket that use
the Server Connection ID using the correct server-facing socket. If the pair is
not unique, the server responds with a CLOSE_SERVER_CID capsule with a CONFLICT
close code. If this occurs, traffic for that Server Connection ID can only use
tunnelled mode, not forwarded.

> If the proxy does not support forwarded mode, or does not allow forwarded mode

Reviewer 2: What's the point of supporting this extension if you don't support forwarding at all?

Author: The other case we describe is using a shared socket from the proxy to the server.

for a particular client or authority by policy, it can reject REGISTER_SERVER_CID
requests with a DENIED close code.

The proxy MUST only forward non-tunnelled packets from the client that are QUIC
short header packets (based on the Header Form bit) and have mapped Server
Connection IDs. Packets sent by the client that are forwarded SHOULD be
considered as activity for restarting QUIC's Idle Timeout {{QUIC}}.

## Removing Mapping State

For any connection ID for which the proxy has sent an acknowledgement, any
mappings for the connection ID last until either endpoint sends a close capsule

> or the entire HTTP request stream closes.

Reviewer 2: QUIC stream states are unidirectional. Should we say that either side of the stream is closed here?

A client that no longer wants a given Connection ID to be forwarded by the
proxy sends a CLOSE_CLIENT_CID or CLOSE_SERVER_CID capsule.

If a client's connection to the proxy is terminated for any reason, all
mappings associated with all requests are removed.

> A proxy can close its server-facing socket once all proxied connections mapped to

Reviewer 2: ```suggestion
A proxy can close its server-facing socket once all UDP proxying requests mapped to
```

# Example

Consider a client that is establishing a new QUIC connection through the proxy.
It has selected a Client Connection ID of 0x31323334. In order to inform a proxy

> of the new QUIC Client Connection ID, the client also sends a sends a

Reviewer: ```suggestion
of the new QUIC Client Connection ID, the client also sends a
```

  Capsule Type = REGISTER_CLIENT_CID
  Connection ID = 0x31323334

DATAGRAM                        -------->
  Quarter Stream ID = 11
  Payload = Encapsulated QUIC initial

           <--------  STREAM(44): HEADERS
                        :status = 200
                        proxy-quic = ?1
                        
           <--------  STREAM(44): DATA
                        Capsule Type = ACK_CLIENT_CID
                        Connection ID = 0x31323334

/* Wait for target server to respond to UDP packet. */

           <--------  DATAGRAM
                        Quarter Stream ID = 11
                        Payload = Encapsulated QUIC initial
~~~

Once the client learns which Connection ID has been selected by the target
server, it can send a new request to the proxy to establish a mapping for
forwarding. In this case, that ID is 0x61626364. The client sends the
following capsule:

~~~
STREAM(44): DATA                -------->
  Capsule Type = REGISTER_SERVER_CID
  Connection ID = 0x61626364
  
           <--------  STREAM(44): DATA
                        Capsule Type = ACK_SERVER_CID
                        Connection ID = 0x61626364
~~~

> Upon receiving a ACK_SERVER_CID capsule, the client starts sending Short Header

Reviewer: ```suggestion
Upon receiving an ACK_SERVER_CID capsule, the client starts sending Short Header
```

# Example

Consider a client that is establishing a new QUIC connection through the proxy.
It has selected a Client Connection ID of 0x31323334. In order to inform a proxy
of the new QUIC Client Connection ID, the client also sends a sends a
REGISTER_CLIENT_CID capsule.

The client will also send the initial QUIC packet with the Long Header form in

> a DATAGRAM frame on the corresponding quarter stream ID.

Reviewer 2: This qualification isn't correct for HTTP 1 and 2

```suggestion
an HTTP datagram.
```

~~~
Client                                             Server

STREAM(44): HEADERS             -------->
  :method = CONNECT
  :protocol = connect-udp
  :scheme = https
  :path = /target.example.com/443/
  :authority = proxy.example.org

STREAM(44): DATA                -------->
  Capsule Type = REGISTER_DATAGRAM_NO_CONTEXT
  Datagram Format Type = UDP_PAYLOAD
  Datagram Format Additional Data = ""
  
STREAM(44): DATA                -------->
  Capsule Type = REGISTER_CLIENT_CID
  Connection ID = 0x31323334

DATAGRAM                        -------->
  Quarter Stream ID = 11
  Payload = Encapsulated QUIC initial

           <--------  STREAM(44): HEADERS
                        :status = 200
                        
           <--------  STREAM(44): DATA
                        Capsule Type = ACK_CLIENT_CID
                        Connection ID = 0x31323334

/* Wait for target server to respond to UDP packet. */

           <--------  DATAGRAM
                        Quarter Stream ID = 11
                        Payload = Encapsulated QUIC initial
~~~

Once the client learns which Connection ID has been selected by the target

> server, it can send a new request to the proxy to establish a mapping for

Reviewer 2: This is making me realize a failure mode:
- client sets up UDP proxying to proxy
- client sends INITIAL with SCID 0x11111111 and DCID 0x33333333
- proxy decapsulates INITIAL and sends it to target server
- server replies but swaps its own CID so it sends INITIAL with DCID 0x11111111 and SCID 0x77777777
- proxy already has a mapping for 0x77777777 on another connection when it receives the server's packet

Author: There is only one mapping that involves the Server CID, which is this:

~~~
(Client-facing socket + Server Connection ID)
    => Server-facing socket
~~~

It only matters to have a conflict about the Server CID within a given client-facing socket (or to the proxy itself).

Connection IDs, or the Connection IDs conflict with other Connection IDs used by
the load balancer, packets can be routed incorrectly.

QUIC-aware proxies that use forwarding mode generally SHOULD NOT be
run behind load balancers; and if they are, they MUST coordinate between the
proxy and the load balancer to create mappings for proxied Connection IDs prior
to the proxy ACK_CLIENT_CID or ACK_SERVER_CID capsules to clients.

QUIC-aware proxies that do not allow forwarding mode can function unmodified
behind QUIC load balancers.

# Packet Size Considerations

Since Initial QUIC packets must be at least 1200 bytes in length, the DATAGRAM

> frames that are used for a QUIC-aware proxy MUST be able to carry at least 1200

Reviewer 2: Replace DATAGRAM frames with HTTP Datagrams everywhere

Author: Replaced!

# IANA Considerations {#iana}

## HTTP Header {#iana-header}

This document registers the "Proxy-QUIC" header in the "Permanent Message
Header Field Names" <[](https://www.iana.org/assignments/message-headers)>.

~~~
    +-------------------+----------+--------+---------------+
    | Header Field Name | Protocol | Status |   Reference   |
    +-------------------+----------+--------+---------------+
    | Proxy-QUIC        |   http   |  exp   | This document |
    +-------------------+----------+--------+---------------+
~~~
{: #iana-header-type-table title="Registered HTTP Header"}

## Capsule Types {#iana-capsule-types}

> This document registers six new values in the in the "HTTP Capsule Types"

Reviewer: ```suggestion
This document registers six new values in the "HTTP Capsule Types"
```

# IANA Considerations {#iana}

## Capsule Types {#iana-capsule-types}

This document registers six new values in the in the "HTTP Capsule Types"
registry established by {{H3DGRAM}}.

|     Capule Type     |   Value   | Specification |
|:--------------------|:----------|:--------------|
| REGISTER_CLIENT_CID | 0xffe100  | This Document |
| REGISTER_SERVER_CID | 0xffe101  | This Document |
| ACK_CLIENT_CID      | 0xffe102  | This Document |
| ACK_SERVER_CID      | 0xffe103  | This Document |
| CLOSE_CLIENT_CID    | 0xffe104  | This Document |
| CLOSE_SERVER_CID    | 0xffe105  | This Document |
{: #iana-format-type-table title="Registered Capsule Type"}

> ## Context Close Codes {#iana-close-codes}

Reviewer 2: ```suggestion
## CID Close Codes {#iana-close-codes}
```

The "HTTP QUIC Connection ID Close Codes" registry governs a 62-bit space.
Registrations in this registry MUST include the following fields:

Type:

: A name or label for the close code.

Value:

: The value of the Close Code field (see {{close-capsule}}) is a 62-bit integer.

Reference:

: An optional reference to a specification for the parameter. This field MAY be
empty.

Registrations follow the "First Come First Served" policy (see Section 4.4 of
{{!IANA-POLICY=RFC8126}}) where two registrations MUST NOT have the same Type
nor Value.

This registry initially contains the following entries:

| Context Close Code   |   Value   | Specification |
|:---------------------|:----------|:--------------|
| NO_ERROR             | 0xff8800  | This Document |
| CONFLICT             | 0xff8801  | This Document |
| DENIED               | 0xff8802  | This Document |
{: #iana-close-codes-table title="QUIC Connection ID Close Code Registry Entries"}

Context close codes with a value of the form 41 * N + 19 for integer values of
N are reserved to exercise the requirement that unknown context close codes be
treated as NO_ERROR. These values MUST NOT be assigned by IANA and MUST NOT

> appear in the listing of assigned values.

Reviewer 2: Now that I've read through the whole thing, I don't think there's value in having CID Close Codes at all. The purpose of Context Close Codes is to allow the client to take different action based on the code. I'm not seeing that here. If we want to indicate "forwarding mode not allowed" I'd suggest using a header instead.

Author: Agreed, I've removed this!
